Resolves bug #550 by fixing the division by zero problem

lehecht opened the pull request:
Perpendicular rectangles were catched in the wrong if-else branch. Fix condition by checking for equal points. Prevent division by zero problem by using alternative formula to compute interception parameter.

lehecht committed:
Resolves bug #550 by fixing the division by zero problem

Perpendicular rectangles were catched in the wrong if-else branch.
Fix condition by checking for equal points. Prevent division by zero
problem by using alternative formula to compute interception parameter.

mzur commented on src/ol/interaction/Draw.js:
```diff
@@ -333,27 +333,25 @@ class Draw extends PointerInteraction {
            */
 
-          if (coordinates.length > 2) {
+           if (coordinates.length > 2) {
```
```suggestion
          if (coordinates.length > 2) {
```

mzur commented on src/ol/interaction/Draw.js:
```diff
@@ -333,27 +333,25 @@ class Draw extends PointerInteraction {
               // going from second in b_vec direction
               // and from third in a_vec direction
-              var x = (third[0] + tmp * (second[1] - third[1]) - second[0]) / (b_vec[0] - b_vec[1] * tmp);
+              var x = (third[0] * second[1] - third[1] * second[0]) / (third[0] * b_vec[1] - third[1] * b_vec[0]);
```
This would now crash if `third` was (0, 0). I tried to do this but I couldn't, so maybe we can ignore this case here.

lehecht replied:
Somehow it's not possible to get (0,0) as third point any more. But I don't know how to rewrite the formula without using `b_vec` and `third` in the denominator. Should we still leave it like this? Another idea would be to add a little bias (e.g. +0.0001), if (0,0) is reached.

mzur replied:
The idea with adding a small value is good! I was about to suggest adding a value all the time but then it could fail if `third` is -0.0001 and the value is +0.0001. I didn't think of a simple check if `third` is 0 :smile: 

I recently saw a similar case where they used `1e-7` as value, so let's randomly use that, too.

mzur commented on src/ol/interaction/Draw.js:
```diff
@@ -333,27 +333,25 @@ class Draw extends PointerInteraction {
+            if (a_vec[0] === 0 && a_vec[1] === 0) {
                 // catch the case where the first and second point are equal
-                coordinates = [[first, first, first, first]];
+                coordinates = [[first]];
```
This still doesn't have the desired effect. I would like this to "reset" the drawing to the state where only a sinlge point was set and the second is floating beneath the cursor. Instead, the rectangle is created with only `first` as single coordinate but internally the `sketchCoords_` still have 3 points and the drawing is finished on the next click. I had a look how to change this but it took me too long. Can you please try to figure this out?

mzur reviewed (commented):
I didn't check the formula but the drawing still works so I assume everything is correct :smile: 

Can you please also apply the same updates to the ellipse shape? The ellipse geometry function looks almost the same than the rectangle one and has the same issues.

lehecht committed:
Remove indention

lehecht committed:
Replace interception parameter formula for ellipse

Prevent division by zero problem by using alternative formula
to compute interception parameter.

lehecht committed:
Fix wrong condition for perpendicular ellipses

lehecht committed:
Add small number to third coordinate

Add bias to prevent division by zero, if variable 'third' is [0,0].

lehecht committed:
Discard duplicate coordinates

mzur reviewed (changes requested):
This works, great! Can you please add comments explaining the special treatment of `third` and the new addition to `addToDrawing_()`? Then it's done!

lehecht committed:
Add comments

mzur committed:
Update comments Draw.js

mzur reviewed (approved):
Thanks! I've tweaked the comments a little so we hopefully understand what's going on when we have to look at this again in two years :wink:
